### ISSUE TITLE
Ability to use link() in afterSave()

### ISSUE THREAD

**Issue by lyhoshva** [yiisoft/yii2#12192]
### What steps will reproduce the problem?

Use method link() in afterSave()

### What is the expected result?

Linked records

### What do you get instead?

Recursive call method afterSave()

### Additional info

| Q                | A
| ---------------- | ---
| Yii version      |  2.0.9
| PHP version      |  5.6.23
| Operating system |  Ubuntu 14.04

**Comment by tunecino** [comment]
It depends on relationship type and on which model are you calling it. If it is a One_to_One or One_to_Many relationship then the link() method of the model itself will internally update the model's `foreingKey` and call its save() method one more time. in such case the afterSave() behavior will be triggered one more time which may cause an infinite loop. If it is the case then it may be easier to just use beforeSave instead by just manually setting the value as save() will be called any way after that:
```php
public function beforeSave($insert)
{
    if ($insert) $this->owner_id = $owner->id;
    return parent::beforeSave($insert);
}
```
for Many_To_Many relationship using method link() inside afterSave() should be easier, I think that is how it is implemented in the[ yii2tech/ar-linkmany](https://github.com/yii2tech/ar-linkmany) extension.

**Comment by lyhoshva** [comment]
I have One_to_One relationship. But in my case i dont know name of id attribute beforehand.

**Comment by lyhoshva** [comment]
@tunecino I have One_to_One relationship. But in my case i dont know name of id attribute beforehand.

**Comment by tunecino** [comment]
did you try an inverse linking within the other model ? I mean instead of `$this->link()` try `$this->owner->link()` that should trigger the other model's behaviors instead. otherwise you may try linking your models at controller's actions, that is how I'd do it. 
also note that github is only for issue tracking or new features suggests. if you think it is a bug and it shouldn't behave that way please try to add more details or a minimal example showing the exact case + data relationship.

**Comment by tunecino** [comment]
another reason to do it at controller action instead is that you don't even need to save that model in the first place. the link() method only require one of them already saved and holding a valid id. it will save the new one for you. check this for more details: http://www.yiiframework.com/doc-2.0/guide-db-active-record.html#saving-relations

**Comment by lyhoshva** [comment]
@tunecino i want hide application logic in model that is why I don't want to use link() in controller's action. I thought that this is a bug because it was fixed in issue #4012 .

**Comment by tunecino** [comment]
@lyhoshva I'm not aware about that issue. I've been in the same place as you are and I solved it changing my app design as it seemed logical error to me. but now I'm not sure if the link() method should have scenarios to prevent recursive calls. I need to learn more about it.
